### ISSUE TITLE
macOS - "Unable to determine BlitzMax installation directory."

### ISSUE THREAD

**Issue by wombatswaffles** [bmx-ng/bmx-ng#43]
Thank you for the new release on macOS! Unfortunately, I'm getting the "Unable to determine BlitzMax installation directory." I have the legacy BlitzMax in "/Applications/BlitzMax" and NG in "/Applications/BlitzMaxNG".

Do I need to do something to actually install it?

**Comment by GWRon** [comment]
This is a MaxIDE issue, not strictly BMX NG

**Comment by woollybah** [comment]
You shouldn't need to do anything other than to unzip it.

MaxIDE should *just run* when you double click it, and realise where it is based on the sub-folders it can see.

I'll do some more tests later from different places on my computer and see if it makes any difference.

Can you provide some information please? Such as your OS version, and what version of XCode you have installed?

Because the app is not signed (and bcc/bmk too), it may not be letting you execute them from MaxIDE. (just a wild guess at the moment). If you could open the terminal and head to the bin folder and try running the following...
```
./bmk -v

./bcc -v
```
They should both output a version line.

**Comment by wombatswaffles** [comment]
It doesn't launch from Finder in any folder.

I am able to launch it from the terminal with "MaxIDE.app/Contents/MacOS/MaxIDE".

macOS version: 10.14
Xcode Version: 10.1 (10B61)

bmk: bmk 3.09 mt-macos-osx-x64 / clang 1000.11.45.5 (cpu x4)
bcc: bcc[ng] Release Version 0.70

**Comment by GWRon** [comment]
As said it is kind of a "maxIDE issue" as maxIDE relies on a "bmx path" which "brl.mod/maxutil.mod" provides. BMK relies on it too .... so might lead to issues.

The real culprit there is, that the function there _relies_ on the existence of only _ONE_ BlitzMax installation (nope, not "portable" then).

Just check this out: (_brl.mod/maxutil.mod/maxutil.bmx_)
```BlitzMax
Function BlitzMaxPath$()
	Global bmxpath$
	If bmxpath Return bmxpath
	Local p$=getenv_("BMXPATH")
	If p
		bmxpath=p
		Return p
	EndIf
	p=AppDir
	Repeat
		Local t$=p+"/bin/bmk"
		?Win32
		t:+".exe"
		?
		If FileType(t)=FILETYPE_FILE
			putenv_ "BMXPATH="+p
			bmxpath=p
			Return p
		EndIf
		Local q$=ExtractDir( p )
		If q=p Throw "Unable to locate BlitzMax path"
		p=q
	Forever
End Function
```
So do you recognize priorities?
- if `BlitzMaxPath` was once run while the app/process is open, use the once evaluated path
- if there is an environment variable "BMXPATH" set, use it
- search "upwards" beginning from "applicationDir/bin"

the function is used somewhere else too (think so at least - because I remember that I had exactly similar issues with "symlinked module folders").


Instead of relying on a "BMXPATH" the function should _onetime_ (per life of the process) find out where it locates the required binaries. "BMXPATH" could be used if the user _desires_ to do so - dunno if it is even needed - for modules or what?
Lookup time _should_ be fast enough to not result in big slowdowns.


@wombatswaffles 
You might open up the file and adjust accordingly:
```
	Local p$=getenv_("BMXPATH")
	If p
		bmxpath=p
		Return p
	EndIf
```
would get removed. Then recompile the modules, recompile "bmk" (from github.com/bmx-ng/bmk.git) and replace it with that, then compile maxIDE and then it should work.
BUT think it is easier to wait for a proper solution by Brucey.

**Comment by woollybah** [comment]
As I kind of surmised, the problem is related to the security stuff in macOS...
I downloaded the archive, unzipped it, and I got a security message the first time I tried to run it.

After allowing the app to run, I got the same "Unable to determine BlitzMax installation directory" message.

It turns out that the app 'PWD' is some random private dir, rather than the bundle dir. This is why MaxIDE is unable to locate bin/bmk, because it has no context from which to find it.

I'm looking into it.

**Comment by GWRon** [comment]
Still it should look for the environment variable...which might lead to the wrong directory (eg. vanilla) ... Or did I misinterpret stuff?

**Comment by wombatswaffles** [comment]
I don't remember it ever being necessary to set an environment variable on macOS for BlitzMax.

**Comment by GWRon** [comment]
How do other "packages" tackle it? With my limited English capabilities and only some minutes of searching all I found were issues people reported with the "app store". I understand that apple tries the users to only download via "app store". But I assume many FOSS projects are only available via github (and the likes). How do they handle it?

A "rebuild everything script" sounds like something people should use. A "one click solution".


Regarding "bundles". So if you created something containing everything then it would work? Talking about something like a "self extracting archive" ... a "binary" having all its stuff incbin'ned and ready to get extracted. On Windows the sandboxing / requires-admin-to-start-first stuff happens only for files having certain "downloaded" flags set. So download it on another computer, put in on an USB stick or copy from a samba share... and voila, ready to execute a binary which you were not allowed to execute if you downloaded them via IE/edge/FF/Chrome...
Maybe something similar is to say about the Apple security: stuff an "allowed" programme creates can get executed with the same rights as this programme... did one try it?
In short words: instead of an archive, provide something like a "setup.exe" (not just an archive which Mac OS handles on its own)
